File: Longest_subarray_with_sum_divisibleby_K.cpp
class Solution{
public:	
	int longSubarrWthSumDivByK(int arr[], int n, int k)
	{
	    // Complete the function
	    map<int,int> lastRec;
	    int prefix=0, maxi=0;
	    for(int i=0 ;i<n;i++){
	        prefix+=arr[i];
	        
	        if(prefix%k==0) maxi=i+1;
	        
	        int rem=prefix%k;
	        if(rem < 0) rem += k;
	        if(lastRec.find(rem)!=lastRec.end()){
	            int subLen= i-(lastRec[rem]);
	            maxi=max(subLen, maxi);
	        }
	        
	        else lastRec[rem]=i;
	    }
	    return maxi;
	}
};

File: Longest_subarray_with_sum_divisibleby_K.js
class Solution {
   
    longSubarrWthSumDivByK(arr,n,k){
        //code here
        const remMap = new Map();
        let prefixSum=0;
        let maxLen=0;
        for(let i=0; i<n;i++){
            prefixSum+=arr[i];
            let rem=prefixSum%k;
            if(rem<0) rem+=k;
            if(rem==0){
                if(i+1>maxLen)
                    maxLen=i+1;
            }
            else{
                //update rem in map
                // let arr=[];
                if(!remMap.has(rem)){
                    remMap.set(rem,i );
                }
                else{
                    if(i-remMap.get(rem)>maxLen){
                        maxLen=i-remMap.get(rem);
                    }
                }
            }
        }
        
        return maxLen;
    }
}